addons/base/views.py

```diff
@@ -210,15 +189,45 @@ def check_access(node, auth, action, cas_resp):
     # `node.is_registration` == True. However, we have no way of telling if
     # `copyfrom` actions are originating from a node being registered.
     # TODO This is raise UNAUTHORIZED for registrations that have not been archived yet
-    if isinstance(node, AbstractNode):
-        if action == 'copyfrom' or (action == 'upload' and node.is_registration):
-            parent = node.parent_node
-            while parent:
-                if parent.can_edit(auth):
-                    return True
-                parent = parent.parent_node
 
-    raise HTTPError(http_status.HTTP_403_FORBIDDEN if auth.user else http_status.HTTP_401_UNAUTHORIZED)
+    required_permission = get_permission_for_action(action)
+    if isinstance(resource, Registration):
+        return _check_registration_permissions(resource, auth, required_permission, action)
+    elif isinstance(resource, Node):
+        return _check_node_permissions(resource, auth, required_permission, action)
+    elif isinstance(resource, Preprint):
+        return _check_preprint_permissions(resource, auth, required_permission)
+    else:
+        raise NotImplementedError()
+
+
+def _check_registration_permissions(registration, auth, permission, action):
+    if permission == permissions.READ:
+        return registration.registered_from.can_view(auth)
+    if action in ('copy_from', 'upload'):
+        return _check_hierarchical_write_permissions(resource=registration, auth=auth)
+    return registration.can_edit(auth)
+
+
+def _check_node_permissions(node, auth, permission, action):
+    if permission == permissions.READ:
+        return node.can_view(auth)
+    return node.can_edit(auth)
```

Comment:
Should be a hierarchical permission check if action is `"copyfrom"`: https://github.com/CenterForOpenScience/osf.io/blob/a990cf2e5d4bfd3cdec63a6674c25fd0951b9c95/addons/base/views.py#L214